From patchwork Tue Oct 20 17:00:55 2020
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Amey Narkhede <ameynarkhede03@gmail.com>
X-Patchwork-Id: 11847445
Return-Path: <SRS0=07X0=D3=vger.kernel.org=kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-12.6 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FORGED_FROMDOMAIN,FREEMAIL_FROM,
	HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_PATCH,MAILING_LIST_MULTI,SIGNED_OFF_BY,
	SPF_HELO_NONE,SPF_PASS,USER_AGENT_GIT autolearn=ham autolearn_force=no
	version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 7D27DC4363A
	for <kvm@archiver.kernel.org>; Tue, 20 Oct 2020 17:01:12 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 1DC1B2222D
	for <kvm@archiver.kernel.org>; Tue, 20 Oct 2020 17:01:10 +0000 (UTC)
Authentication-Results: mail.kernel.org;
	dkim=pass (2048-bit key) header.d=gmail.com header.i=@gmail.com
 header.b="tXJEuhUh"
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S2395096AbgJTRBJ (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Tue, 20 Oct 2020 13:01:09 -0400
Received: from lindbergh.monkeyblade.net ([23.128.96.19]:35596 "EHLO
        lindbergh.monkeyblade.net" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S2395094AbgJTRBJ (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 20 Oct 2020 13:01:09 -0400
Received: from mail-pj1-x1043.google.com (mail-pj1-x1043.google.com
 [IPv6:2607:f8b0:4864:20::1043])
        by lindbergh.monkeyblade.net (Postfix) with ESMTPS id 9EB5EC0613CE
        for <kvm@vger.kernel.org>; Tue, 20 Oct 2020 10:01:07 -0700 (PDT)
Received: by mail-pj1-x1043.google.com with SMTP id a17so1207882pju.1
        for <kvm@vger.kernel.org>; Tue, 20 Oct 2020 10:01:07 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id:in-reply-to:references
         :mime-version:content-transfer-encoding;
        bh=eaYqmjcMx802IYh8As8yQQI4HZLzXd0/F879o5VXjTg=;
        b=tXJEuhUh/jBCkoIm18hk/J+QQErGA6dM6aodvpm5zAjpd0SMaC9cj5NzSsVeomWwdb
         uMISMjUWoDPDuNAFVKlYOEfjlphlNV77EnuuF/VObOBaIsCnq0kxkm1iqywQmpNaxH7s
         RakE6JuJh9Yeru7KCeiEDvMZlJ1ZHkPPdjt2auaANrVz8k8PotKbJibCkHe+VM/CQG/t
         YyalaEC9NvE8uiwSmit9I57v0mAWMyknlHFF5bNd9XGOeEJ/whGazNu0uiM/t4PgAyvK
         i+LjYHhxFKarkk6JTNrJKpMFGFJotDXrai7JIvgo6GXWnYfoAmY+DfG4gmom9gteZ5Xu
         uuQA==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=eaYqmjcMx802IYh8As8yQQI4HZLzXd0/F879o5VXjTg=;
        b=Q/VUMq3D67vb+64j3zh46UUvRQtCq9Diqhs9S+tW+PK3KwXVVLr6iFpdtKzg0oqHuo
         19DITorkmpjs+5W4jUdG04UNrn7RKgYwDdgBnlmD47w0qRi1YK1dPSkay8lpbbpTHQ4f
         xgQhOoxVx44APowa/hoqELqDO7T//aKhu68cCKxPMym2BJVsgRkSGHSRKrZXWZf9euYn
         ii6Qo00yfNK2W/KeqwNCQpkzgwHFDiYZ26HtTxOS8f2czgfdklFrmyC8gNSDaNrYhS+R
         J+oX7SdXG6NuKrBGTUzwrHD9VjP1WXy12WnI9hMAgfgpD0mU9dy7GTRl6hhdlf+xs6CF
         kaMw==
X-Gm-Message-State: AOAM530CIHC2FhebHVsxwHBUm1Yv04uoNUW7ytrRjDmWGF7AJS/u1mtR
        QAa3MuknRMyvGbIG+VZHhmk=
X-Google-Smtp-Source: 
 ABdhPJyrIZ0jrQpUiVQiMr6asT4wMDG/t4eFF26d8JnKNK5AR8/YFD8r+JrSG6otU6RLsyXjrWbfiQ==
X-Received: by 2002:a17:902:c24b:b029:d5:f14a:318d with SMTP id
 11-20020a170902c24bb02900d5f14a318dmr932875plg.4.1603213267195;
        Tue, 20 Oct 2020 10:01:07 -0700 (PDT)
Received: from localhost.localdomain ([103.248.31.132])
        by smtp.googlemail.com with ESMTPSA id
 x29sm2766161pfp.152.2020.10.20.10.01.03
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Tue, 20 Oct 2020 10:01:06 -0700 (PDT)
From: Amey Narkhede <ameynarkhede03@gmail.com>
To: qemu-devel@nongnu.org, ameynarkhede03@gmail.com
Cc: "Michael S. Tsirkin" <mst@redhat.com>,
        Cornelia Huck <cohuck@redhat.com>, kvm@vger.kernel.org,
        Paolo Bonzini <pbonzini@redhat.com>
Subject: [PATCH 1/2] linux-headers: Add support for reads in ioeventfd
Date: Tue, 20 Oct 2020 22:30:55 +0530
Message-Id: <20201020170056.433528-2-ameynarkhede03@gmail.com>
X-Mailer: git-send-email 2.28.0
In-Reply-To: <20201020170056.433528-1-ameynarkhede03@gmail.com>
References: <20201020170056.433528-1-ameynarkhede03@gmail.com>
MIME-Version: 1.0
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

This patch introduces a new flag KVM_IOEVENTFD_FLAG_DATAREAD
in ioeventfd to enable receiving a notification when a
guest reads from registered PIO/MMIO address.

Signed-off-by: Amey Narkhede <ameynarkhede03@gmail.com>
---
 linux-headers/linux/kvm.h | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/linux-headers/linux/kvm.h b/linux-headers/linux/kvm.h
index 43580c767c..3e71d15a53 100644
--- a/linux-headers/linux/kvm.h
+++ b/linux-headers/linux/kvm.h
@@ -695,6 +695,7 @@ struct kvm_guest_debug {
 
 enum {
 	kvm_ioeventfd_flag_nr_datamatch,
+	kvm_ioeventfd_flag_nr_dataread,
 	kvm_ioeventfd_flag_nr_pio,
 	kvm_ioeventfd_flag_nr_deassign,
 	kvm_ioeventfd_flag_nr_virtio_ccw_notify,
@@ -703,6 +704,7 @@ enum {
 };
 
 #define KVM_IOEVENTFD_FLAG_DATAMATCH (1 << kvm_ioeventfd_flag_nr_datamatch)
+#define KVM_IOEVENTFD_FLAG_DATAREAD (1 << kvm_ioeventfd_flag_nr_dataread)
 #define KVM_IOEVENTFD_FLAG_PIO       (1 << kvm_ioeventfd_flag_nr_pio)
 #define KVM_IOEVENTFD_FLAG_DEASSIGN  (1 << kvm_ioeventfd_flag_nr_deassign)
 #define KVM_IOEVENTFD_FLAG_VIRTIO_CCW_NOTIFY \
@@ -712,11 +714,12 @@ enum {
 
 struct kvm_ioeventfd {
 	__u64 datamatch;
+	__u64 dataread;
 	__u64 addr;        /* legal pio/mmio address */
 	__u32 len;         /* 1, 2, 4, or 8 bytes; or 0 to ignore length */
 	__s32 fd;
 	__u32 flags;
-	__u8  pad[36];
+	__u8  pad[28];
 };
 
 #define KVM_X86_DISABLE_EXITS_MWAIT          (1 << 0)

From patchwork Tue Oct 20 17:00:56 2020
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Amey Narkhede <ameynarkhede03@gmail.com>
X-Patchwork-Id: 11847447
Return-Path: <SRS0=07X0=D3=vger.kernel.org=kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-12.6 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FORGED_FROMDOMAIN,FREEMAIL_FROM,
	HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_PATCH,MAILING_LIST_MULTI,SIGNED_OFF_BY,
	SPF_HELO_NONE,SPF_PASS,USER_AGENT_GIT autolearn=ham autolearn_force=no
	version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id C9202C388F2
	for <kvm@archiver.kernel.org>; Tue, 20 Oct 2020 17:01:12 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 5F5F32177B
	for <kvm@archiver.kernel.org>; Tue, 20 Oct 2020 17:01:12 +0000 (UTC)
Authentication-Results: mail.kernel.org;
	dkim=pass (2048-bit key) header.d=gmail.com header.i=@gmail.com
 header.b="c4SWT7SE"
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S2395098AbgJTRBL (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Tue, 20 Oct 2020 13:01:11 -0400
Received: from lindbergh.monkeyblade.net ([23.128.96.19]:35608 "EHLO
        lindbergh.monkeyblade.net" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S2395094AbgJTRBK (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 20 Oct 2020 13:01:10 -0400
Received: from mail-pl1-x643.google.com (mail-pl1-x643.google.com
 [IPv6:2607:f8b0:4864:20::643])
        by lindbergh.monkeyblade.net (Postfix) with ESMTPS id 65653C0613CE
        for <kvm@vger.kernel.org>; Tue, 20 Oct 2020 10:01:10 -0700 (PDT)
Received: by mail-pl1-x643.google.com with SMTP id h2so1325325pll.11
        for <kvm@vger.kernel.org>; Tue, 20 Oct 2020 10:01:10 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id:in-reply-to:references
         :mime-version:content-transfer-encoding;
        bh=9vfxdnRCLsliVllqt/rhZ7KYxjMhIVE68F6rGYVRGI8=;
        b=c4SWT7SE6bL+6PNYDzpJ6TlgbtT73FuFzyvk9J1igWT3KUmasgTurJVm/AIkw3em+h
         MIwe8tfpZu/kigbbr3k5Vct5Pkl65y+DGZZLCmueW5xKQMFgiojz2BupFYiqNGXkl3eB
         bra6py56A4YVL8MTB8KkXcwYVVyRF5CV+nYPbg9D/1ium83fI7UrGIlQ6ZrPfUNxrvYX
         xWSCsaetiSB05dKeeNJoLcZnfxzy1fmVnqG31tqd/0XphHsrGDYeACt0PbJ9V6pkJ3Yc
         TwXKBmaqKi1pkKaoF/NAWDlXsu19WM1H6BooydwZ4tSA2HSpfY2cWZsk+jpsEz/8HlWD
         NUNg==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=9vfxdnRCLsliVllqt/rhZ7KYxjMhIVE68F6rGYVRGI8=;
        b=omrD7sJ54o6nn2wE4c1C6JGySFd6ibQqj67DKUCUiDa9ryhCPanQCM7DE0dpQuDLrI
         gDGHUF06lw2wVb79EBRt/vnVq9HbUYcgA2GU8UwB+a5kFd3OaJDkhjhkF5ZTvU9u3NQo
         KhaYUJwgYt5Au7Rx0bkpMo7ZoKGPxBmasorFeeuSFyEZRKQcLa9GupTccu6Ceq/f02BG
         8BXEEXby1odtj7HrBbHmJfC90la+oqONgLGqpo1EvFzYb+n51YJCI6/xSEqdzrzlksRy
         scg/kXk1vTE1tp65C1pQZnp+X3lx5TICfmFZQ2S2lVAd+2cdiHndHo/TjwfprIdA0P5q
         L8og==
X-Gm-Message-State: AOAM531jMmfNQkUdYsenbwQZESGvnFZeuKTTKL4UUh3RIwbAqPzMqRZf
        WClFi29dBjP+q0u/5lpSL7seUianbL0=
X-Google-Smtp-Source: 
 ABdhPJwiTW478eHb035jsmIY5P9juzumk5Lm7S0w1b/UrmSeJgsmdmd/ZaKLAyxhaadizHpiCeJt9g==
X-Received: by 2002:a17:90a:bb0e:: with SMTP id
 u14mr3608130pjr.112.1603213269952;
        Tue, 20 Oct 2020 10:01:09 -0700 (PDT)
Received: from localhost.localdomain ([103.248.31.132])
        by smtp.googlemail.com with ESMTPSA id
 x29sm2766161pfp.152.2020.10.20.10.01.07
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Tue, 20 Oct 2020 10:01:09 -0700 (PDT)
From: Amey Narkhede <ameynarkhede03@gmail.com>
To: qemu-devel@nongnu.org, ameynarkhede03@gmail.com
Cc: "Michael S. Tsirkin" <mst@redhat.com>,
        Cornelia Huck <cohuck@redhat.com>, kvm@vger.kernel.org,
        Paolo Bonzini <pbonzini@redhat.com>
Subject: [PATCH 2/2] kvm: Add ioeventfd read test code
Date: Tue, 20 Oct 2020 22:30:56 +0530
Message-Id: <20201020170056.433528-3-ameynarkhede03@gmail.com>
X-Mailer: git-send-email 2.28.0
In-Reply-To: <20201020170056.433528-1-ameynarkhede03@gmail.com>
References: <20201020170056.433528-1-ameynarkhede03@gmail.com>
MIME-Version: 1.0
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

This patch adds kvm_set_ioeventfd_read and
dummy_notifier_read functons to test ioeventfd
read support. When the guess writes to address
provided in kvm_set_ioeventfd_read function,
dummy_notifier_read prints to stdio.

Signed-off-by: Amey Narkhede <ameynarkhede03@gmail.com>
---
 accel/kvm/kvm-all.c | 55 +++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 55 insertions(+)

diff --git a/accel/kvm/kvm-all.c b/accel/kvm/kvm-all.c
index 9ef5daf4c5..357e74d84c 100644
--- a/accel/kvm/kvm-all.c
+++ b/accel/kvm/kvm-all.c
@@ -1006,6 +1006,43 @@ static uint32_t adjust_ioeventfd_endianness(uint32_t val, uint32_t size)
     return val;
 }
 
+static void dummy_notifier_read(EventNotifier *n)
+{
+    printf("Received ioeventfd read event\n");
+    event_notifier_test_and_clear(n);
+}
+
+static int kvm_set_ioeventfd_read(int fd, hwaddr addr, uint64_t val,
+				  uint64_t size, bool datamatch)
+{
+	int ret;
+	struct kvm_ioeventfd ioevent = {
+        	.datamatch = datamatch ? adjust_ioeventfd_endianness(val, size) : 0,
+		.dataread = val,
+		.addr = addr,
+		.len = size,
+		.flags = KVM_IOEVENTFD_FLAG_DATAREAD,
+		.fd = fd,
+	};
+
+	if (!kvm_enabled()) {
+		return -ENOSYS;
+	}
+
+	if (datamatch) {
+            ioevent.flags |= KVM_IOEVENTFD_FLAG_DATAMATCH;
+	}
+
+	ret = kvm_vm_ioctl(kvm_state, KVM_IOEVENTFD, &ioevent);
+
+	if (ret < 0) {
+		return -errno;
+	}
+
+	return 0;
+}
+
+
 static int kvm_set_ioeventfd_mmio(int fd, hwaddr addr, uint32_t val,
                                   bool assign, uint32_t size, bool datamatch)
 {
@@ -2012,6 +2049,7 @@ static int kvm_init(MachineState *ms)
     KVMState *s;
     const KVMCapabilityInfo *missing_cap;
     int ret;
+    int efd = -1;
     int type = 0;
     const char *kvm_type;
     uint64_t dirty_log_manual_caps;
@@ -2253,6 +2291,22 @@ static int kvm_init(MachineState *ms)
     }
 
     cpus_register_accel(&kvm_cpus);
+
+    EventNotifier *e = g_malloc0(sizeof(EventNotifier));
+    ret = event_notifier_init(e, false);
+    if (ret < 0) {
+	printf("Failed to initialize EventNotifier\n");
+    }
+    else {
+        AioContext *ctx = qemu_get_aio_context();
+        efd = event_notifier_get_fd(e);
+        aio_set_event_notifier(ctx, e, false, dummy_notifier_read, NULL);
+        ret = kvm_set_ioeventfd_read(efd, 0xff01003f, 123, 8, false);
+        if (ret < 0)
+            printf("ioeventfd read failed\n");
+    }
+
+
     return 0;
 
 err:
@@ -2268,6 +2322,7 @@ err:
     return ret;
 }
 
+
 void kvm_set_sigmask_len(KVMState *s, unsigned int sigmask_len)
 {
     s->sigmask_len = sigmask_len;
